File: packages/cli/tests/docs-cli-drift.e2e.test.ts
import { readdir, readFile } from "node:fs/promises";
import { join, relative, resolve } from "node:path";

import { describe, expect, it } from "vitest";

const workspaceRoot = resolve(import.meta.dirname, "../../..");
const docsRoot = join(
  workspaceRoot,
  "packages",
  "homepage",
  "src",
  "content",
  "docs",
);

const collectContentFiles = async (directory: string): Promise<string[]> => {
  const entries = await readdir(directory, { withFileTypes: true });
  const files = await Promise.all(
    entries.map(async (entry) => {
      const fullPath = join(directory, entry.name);

      if (entry.isDirectory()) {
        return collectContentFiles(fullPath);
      }

      return entry.isFile() && /\.(md|mdx)$/u.test(entry.name)
        ? [fullPath]
        : [];
    }),
  );

  return files.flat();
};

const unsupportedInitFlags = [
  {
    name: "--key",
    pattern: /(^|[\s`])--key(?![-\w])/u,
  },
  {
    name: "--promptKey",
    pattern: /--promptKey/u,
  },
  {
    name: "--url",
    pattern: /--url\b/u,
  },
] as const;

const staleSecretArtifactFragments = [
  {
    name: "`.age` secret artifact suffix",
    pattern: /`\.age`/u,
  },
  {
    name: ".age secret artifact example",
    pattern: /\b(?:config|credentials|work-key)\.age\b/u,
  },
] as const;

describe("documentation CLI drift", () => {
  it("does not document removed init credential flags", async () => {
    const files = [
      join(workspaceRoot, "README.md"),
      ...(await collectContentFiles(docsRoot)),
    ];
    const violations: string[] = [];

    for (const file of files) {
      const text = await readFile(file, "utf8");
      const lines = text.split(/\r?\n/u);

      for (const [index, line] of lines.entries()) {
        for (const flag of unsupportedInitFlags) {
          if (flag.pattern.test(line)) {
            violations.push(
              `${relative(workspaceRoot, file)}:${index + 1} uses ${flag.name}`,
            );
          }
        }
      }
    }

    expect(violations).toEqual([]);
  });

  it("does not document the legacy .age secret artifact suffix", async () => {
    const files = await collectContentFiles(docsRoot);
    const violations: string[] = [];

    for (const file of files) {
      const text = await readFile(file, "utf8");
      const lines = text.split(/\r?\n/u);

      for (const [index, line] of lines.entries()) {
        for (const fragment of staleSecretArtifactFragments) {
          if (fragment.pattern.test(line)) {
            violations.push(
              `${relative(workspaceRoot, file)}:${index + 1} uses ${fragment.name}`,
            );
          }
        }
      }
    }

    expect(violations).toEqual([]);
  });
});

File: test/e2e/docs_cli_drift_e2e_test.dart
// Dart port of `tests/docs-cli-drift.e2e.test.ts`.
//
// The dotweave package root is the repo root, so the homepage docs paths
// resolve from it directly, regardless of the test runner's working directory.

@Timeout(Duration(minutes: 3))
library;

import 'dart:io';

import 'package:path/path.dart' as p;
import 'package:test/test.dart';

import '../helpers/e2e_context.dart';

final String _workspaceRoot = e2ePackageRoot();
final String _docsRoot = p.join(_workspaceRoot, 'homepage', 'app', 'content');

final RegExp _contentFilePattern = RegExp(r'\.(md|mdx)$');

Future<List<String>> _collectContentFiles(String directory) async {
  final files = <String>[];

  await for (final entry in Directory(directory).list(followLinks: false)) {
    if (entry is Directory) {
      files.addAll(await _collectContentFiles(entry.path));
    } else if (entry is File &&
        _contentFilePattern.hasMatch(p.basename(entry.path))) {
      files.add(entry.path);
    }
  }

  return files;
}

final List<({String name, RegExp pattern})> _unsupportedInitFlags = [
  (name: '--key', pattern: RegExp(r'(^|[\s`])--key(?![-\w])')),
  (name: '--promptKey', pattern: RegExp('--promptKey')),
  (name: '--url', pattern: RegExp(r'--url\b')),
];

final List<({String name, RegExp pattern})> _staleSecretArtifactFragments = [
  (name: '`.age` secret artifact suffix', pattern: RegExp(r'`\.age`')),
  (
    name: '.age secret artifact example',
    pattern: RegExp(r'\b(?:config|credentials|work-key)\.age\b'),
  ),
];

final RegExp _linePattern = RegExp(r'\r?\n');

void main() {
  group('documentation CLI drift', () {
    test('does not document removed init credential flags', () async {
      final files = [
        p.join(_workspaceRoot, 'README.md'),
        ...await _collectContentFiles(_docsRoot),
      ];
      final violations = <String>[];

      for (final file in files) {
        final text = await File(file).readAsString();
        final lines = text.split(_linePattern);

        for (var index = 0; index < lines.length; index += 1) {
          final line = lines[index];

          for (final flag in _unsupportedInitFlags) {
            if (flag.pattern.hasMatch(line)) {
              violations.add(
                '${p.relative(file, from: _workspaceRoot)}:${index + 1} '
                'uses ${flag.name}',
              );
            }
          }
        }
      }

      expect(violations, isEmpty);
    });

    test('does not document the legacy .age secret artifact suffix', () async {
      final files = await _collectContentFiles(_docsRoot);
      final violations = <String>[];

      for (final file in files) {
        final text = await File(file).readAsString();
        final lines = text.split(_linePattern);

        for (var index = 0; index < lines.length; index += 1) {
          final line = lines[index];

          for (final fragment in _staleSecretArtifactFragments) {
            if (fragment.pattern.hasMatch(line)) {
              violations.add(
                '${p.relative(file, from: _workspaceRoot)}:${index + 1} '
                'uses ${fragment.name}',
              );
            }
          }
        }
      }

      expect(violations, isEmpty);
    });
  });
}
